alf/layers.py

```diff
@@ -705,6 +706,139 @@ def bias(self):
+        ``[B, C', L]``, by causal we mean ``y[..., l]`` only depends on
+        ``X[..., :l]`` (i.e. the past), and there is no dependency on
+        ``X[..., l:]`` (i.e. future) as in the standard non-causal
+        convolution.
+
+        This can implemented by using an asymmetric padding, which in effect
+        shift the input to the right (future) according to kernel size.
+
+        Args:
+            in_channels (int): channels of the input
+            out_channels (int): channels of the output
+            kernel_size (int): size of the kernel
+            dilation (int): controls the spacing between the kernel points.
+                Please refer to here for a visual illustration:
+                https://github.com/vdumoulin/conv_arithmetic/blob/master/README.md
+            hide_current (bool): whether to hide the current by shifting the
+                input to the right (future) by one. This is typically needed
+                in the first layer of a causal conv net.
+            activation (torch.nn.functional): activation to be applied to output
+            use_bias (bool|None): whether use bias. If None, will use ``not use_bn``
+            use_bn (bool): whether use batch normalization
+            kernel_initializer (Callable): initializer for the conv layer kernel.
+                If None is provided a variance_scaling_initializer with gain as
+                ``kernel_init_gain`` will be used.
+            kernel_init_gain (float): a scaling factor (gain) applied to the
+                std of kernel init distribution. It will be ignored if
+                ``kernel_initializer`` is not None.
+            bias_init_value (float): a constant
+        """
+        super(CausalConv1D, self).__init__()
+        if use_bias is None:
+            use_bias = not use_bn
+        self._activation = activation
+
+        # use F.pad for asymmetric padding
+        if hide_current:
+            assert dilation == 1, "the dilation should be 1 for hiding current"
+            asymmetric_padding = (kernel_size, -1)
+        else:
+            asymmetric_padding = ((kernel_size - 1) * dilation, 0)
+
+        self._pad = partial(
+            F.pad, pad=asymmetric_padding, mode='constant', value=0)
+        self._causal_conv1d = nn.Conv1d(
+            in_channels,
+            out_channels,
+            kernel_size,
+            stride=1,
+            padding=0,
+            dilation=dilation,
+            bias=use_bias)
+
+        self._kernel_initializer = kernel_initializer
+        self._kernel_init_gain = kernel_init_gain
+        self._bias_init_value = bias_init_value
+        self._use_bias = use_bias
+        if use_bn:
+            self._bn = nn.BatchNorm1d(out_channels)
+        else:
+            self._bn = None
+
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        """Initialize the parameters."""
+        if self._kernel_initializer is None:
+            variance_scaling_init(
```

Comment:
One last minor potential issue: variance_scaling_init calculates the gain by assuming the input is N(0,1). Now since we know we are always padding the input with zeros, this auto calculation might be problematic.